File: 프로그래머스/3/42895. N으로 표현/N으로 표현.java
/*
- N을 몇 번 써서 number를 만들 수 있는가?
- 사칙연산 사용 가능, 괄호 ok
- N은 8번까지만 사용 가능
- 최솟값 구하기 -> 최적화 문제 ->조건: n을 몇번 썼는가로 탐색 
- dp[k] = N을 K번 사용해서 만들 수 있는 숫자들의 집합
- N을 여러 번 써서 뭔가를 만든다 → "횟수 기반 DP"
*/
import java.util.*;

class Solution {
    public int solution(int N, int number) {
        if (N == number) return 1;
        
        List<Set<Integer>> dp = new ArrayList<>();
        // dp[0]은 사용하지 않고 1~8만 사용
        for(int i = 0; i <= 8; i++){
            dp.add(new HashSet<>());
        }
        
        for(int i = 1; i <= 8; i++){
            // 같은 숫자 반복해서 만든 수 추가 (5, 55, 555)
            int repeatedNum = Integer.parseInt(String.valueOf(N).repeat(i));
            dp.get(i).add(repeatedNum);
            
            // 조합을 통해 가능한 값 추가
            for(int j = 1; j < i; j++){ // i 2개 / j 1개 / i-j는 i개 이외의 나머지에서 가져오겠다는 뜻
                for(int op1 : dp.get(j)){
                    for(int op2 : dp.get(i-j)){
                        dp.get(i).add(op1 + op2);
                        dp.get(i).add(op1 - op2);
                        dp.get(i).add(op1 * op2);
                        if(op2 != 0) dp.get(i).add(op1 / op2);
                    }
                }
            }
            if(dp.get(i).contains(number)) return i;
        }
        return -1;
    }
}

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.py
# N을 i번 사용해서 만들 수 있는 수의 집합을 dp[i]에 저장
# dp[1] = {5}
# dp[2] = {55, 5+5, 5-5, 5*5, 5//5}
# dp[3] = {555, (55+5), (5+55), (5+5+5), ...}

# dp[i]를 구할 때는 1<= j <i에 대해서 dp[j]와 dp[i-j]의 사칙연산 조합 모두 시도
# 4개로 만든다고 하면 2개 + 2개 , 3개 + 1개 이런식으로 조합
# 모든 경우의 수 set에 저장
# i= 1~8까지 돌면서 number 있으면 그 i가 최소 횟수 됨
def solution(N, number):
    # 초기 세팅
    dp = [set() for _ in range(9)] # 8번까지 사용 가능 dp[1] ~ dp[8]
    
    for i in range(1, 9):
        # 1. N을 i번 이어 붙인 수 추가 (5, 55, 555 ...)
        dp[i].add(int(str(N) * i))
    
        # 2. 사칙연산으로 만들 수 있는 모든 수 추가
        for j in range(1, i):
            for op1 in dp[j]:
                for op2 in dp[i-j]:
                    dp[i].add(op1 + op2)
                    dp[i].add(op1 - op2)
                    dp[i].add(op1 * op2)
                    if op2 != 0:
                        dp[i].add(op1//op2)
        
        # 3. 현재 i개의 N으로 number를 만들 수 있다면 바로 return
        if number in dp[i]:
            return i
    
    return -1 # 8개 이하로 못만들면 -11